[8.0][ADD] product_variant_configurator

ThomasBinsfeld opened the pull request:
- [x] README
- [ ] Unit tests

oihane committed:
[MOD] New name and security files fixed

oihane committed:
[MOD] <product_variants_no_automatic_creation> _product_find

oihane committed:
[MOD] Fix pep8

oihane committed:
[MOD] possible values added to scheduled lines for attributes

oihane committed:
[MOD] Attributes in mo product lines

oihane committed:
[MOD] Operation assignation fixing

oihane committed:
[MOD] Fixed errors and changes

oihane committed:
[MOD] translations

oihane committed:
[MOD] get inherit attribute value, and corresponding product if exists

oihane committed:
[MOD] <product_variants_no_automatic_creation> prices according to attributes

oihane committed:
[MOD] Fixing pricelist aplication

oihane committed:
[FIX] PEP8

oihane committed:
[MOD] import missing

oihane committed:
[FIX] <product_variants_no_automatic_creation> view don't allow editing

oihane committed:
Update product_view.xml

pedrobaeza committed:
[IMP] product_variants_no_automatic_creation: Enhance labels and defaults

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Labels

oihane committed:
[MOD] variants creation fixed

oihane committed:
[MOD] Avoid launching onchange with default value

oihane committed:
[MOD] Sale_variants description changed, and uom in purchase when template selected

pedrobaeza committed:
[IMP] Expand AUTHORS

pedrobaeza committed:
[FIX/IMP] product_variants_no_automatic_creation: Preserve order on attributes + improve code

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Make product template search usable without variants. Closes #804

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Adapt _product_find for looking for non full-set of values

sysadminmatmoz committed:
Missing templates and translations added

pedrobaeza committed:
[IMP] product_variants_no_automatic_creation: Include constraint for product/attribute uniqueness

pedrobaeza committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Find correctly a product for its attributes

Adding to the domain all values that exists, not only the ones that have more than one
possible value. That was an old constraint because attributes with only one value were
not populated, but that was fixed a long time ago and now it makes that the match doesn't
happen because the comparison of the attributes length fail.

pedrobaeza committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] product_variants_no_automatic_creation: Add configurator in product variant form

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Security for product.product.attribute

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Don't error on missing attribute values

pedrobaeza committed:
[IMP] product_variants_no_automatic_creation: Avoid to disturb tests results

pedrobaeza committed:
[FIX] product_variants_no_automatic_creation: Fix security file

jcarlosmontoya committed:
Update term es.po

pedrobaeza committed:
[IMP] product_variant_*: Abstract product configurator + application on stock.move

hugosantosred committed:
Fix _check_configuration_validity constraint

oihane committed:
[FIX+IMP] product_variants_no_automatic_creation:

* Fix expected singleton error when updating attributes (#1237)
* Tests

pedrobaeza committed:
[IMP] product_variants_no_automatic_creation

* Link quick-created attribute value to the template
* Don't require parent product_tmpl_id field in view

ThomasBinsfeld commented:
@pedrobaeza

pedrobaeza commented:
You should respect commit history following a little variation of this guide and then make the renaming: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-9.0#technical-method-to-migrate-a-module-from-80-to-90-branch

zakiuu commented:
@pedrobaeza @sbidoul  Hi,
I've been trying working on the test coverage for the product_variant_configurator, and I could get it to almost 80%, Although there are some cases that can be only tested with other modules such as purchase, sales, MRP.
My target is to add the product configurator in the purchase module, where the user can select a product template, and gives values to the attributes related to the selected template in the purchase line. When saving the purchase order the 'product.product' gets created on the fly.
I've checked in 'odoomrp' repo that such module exists. My question is: 
should I port the 'odoomrp' module to OCA in this PR or create a new one?

Garamotte commented:
@zakiuu : Please also look at my review on the 9.0 PR : https://github.com/OCA/product-variant/pull/39#pullrequestreview-3411235
Some comments may be applicable to this PR.

zakiuu commented:
@sylvain-garancher Hi,
Can you please add more reviews here. I did take care of the reviews you've put in #39

Garamotte commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -0,0 +1,245 @@
+                    'owner_model': self._name,
+                    'owner_id': self.id,
+                })
+            self.product_attribute_ids = [(0, 0, x) for x in attribute_list]
```
You can directly append the tuples `(0, 0, data)` in `attribute_list` and remove the for loop.

Garamotte commented on product_variant_configurator/models/product_configurator.py:
```diff
@@ -0,0 +1,245 @@
+                obj = self.env['product.attribute.value'].with_context(
+                    lang=self.partner_id.lang)
+                values = obj.browse(
+                    self.product_attribute_ids.mapped('value_id').ids)
```
You should use `values.ids`, instead of calling the `mapped` method twice.

Garamotte commented on product_variant_configurator/models/pricelist.py:
```diff
@@ -0,0 +1,161 @@
+            'ORDER BY sequence, min_quantity desc',
+            (prod_tmpl_ids, categ_ids, version.id))
+        item_ids = [x[0] for x in cr.fetchall()]
+        items = self.env['product.pricelist.item'].browse(item_ids)
```
Maybe we should use `items = self.env['product.pricelist.item'].browse(item_ids).exists()`, to be sure we only read available ids for the current user, because the SQL query doesn't check access rules (`ir.rule`).

@pedrobaeza : What's your opinion about that ?

pedrobaeza replied:
This was copied directly from Odoo code, so it should be the same. That's the part that I like less, because I didn't want to override full method. Maybe we can look for a better implementation of this part.

zakiuu replied:
@pedrobaeza @sylvain-garancher I did spend a lot of time looking at this method, even the code wasn't fully copied. I checked how it is done in v10 and have adapted to v8 but when I did unit tests I got some bugs, 
I can push it here if you want to have a look at it

pedrobaeza replied:
Well, we can take a look.

zakiuu replied:
@pedrobaeza what is the use of overriding the pricelist method here in this module?

pedrobaeza replied:
The main purpose is to calculate the correct price when yo still don't have a product variant created, but the rules for product template and product categories still apply.

rvalyi replied:
this method is more than 130 lines... I think it's a bit too much... Isn't possible to break it down into smaller pieces that would make sense for potential overriders and make it easier to read?

zakiuu replied:
I don't believe this is the right implementation of the behavior that we want to have. 
I would take it out from this module and have a  specific module named price_list product configurator where a rule can be defined using the product attribute configurator, and give the user the option of adding the value of the extra price to the base price.
to resume:
- Add one2many product attribute configurator field to the pricelist item object.
- Add boolean field  'include extra price to base' to take care of the extra price that the variant might have.
- override the _price_rule_get_multi to include the new behavior.
  What do you think guys
  @sbidoul

pedrobaeza replied:
This method was a copy/paste from the core one. The idea was to ease the identification of differences from upstream.  Maybe you can find a way to inherit correctly without doing this hard copy. @oihane can comment about the problems she found.

rvalyi replied:
@pedrobaeza got it. In this case indeed this is acceptable indeed. Such a comment in the code would be welcome.

pedrobaeza commented:
@zakiuu The preservation of the commit history is still needed in the PR.

Garamotte commented:
I just had a traceback when testing on Runbot :

```
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3269415-38-07fea9/openerp/addons/product_variant_configurator/models/product_configurator.py", line 45, in onchange_product_tmpl_id
    self.product_id = self.product_tmpl_id.product_variant_ids
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3269415-38-07fea9/openerp/fields.py", line 851, in __set__
    value = self.convert_to_cache(value, record)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3269415-38-07fea9/openerp/fields.py", line 1581, in convert_to_cache
    raise ValueError("Wrong value for %s: %r" % (self, value))
ValueError: Wrong value for product.product.product_id: product.product(41, <openerp.models.NewId object at 0x7f53b9347510>)
```

Test done on this instance : https://runbot.odoo-community.org/runbot/build/3269415

Steps to reproduce : Go to the `product.product` form view, click `Create`, select a value in the `product_tmpl_id` field.

Probable cause : I was trying to create a variant for a product which already has a variant, but no attributes defined (Blank CD, in my test)

ghost commented:
I was also able to reproduce the error.

It occurs when creating a new Product Variant and using a product that has only 1 variant (i.e. has an empty _Variants_ tab) as template.

pedrobaeza commented:
@ThomasBinsfeld @zakiuu, are you going to continue this PR?

zakiuu commented:
Yeah ofcorse we were juste busy with a project release

On Thu, Oct 20, 2016, 01:05 Pedro M. Baeza notifications@github.com wrote:

> @ThomasBinsfeld https://github.com/ThomasBinsfeld @zakiuu
> https://github.com/zakiuu, are you going to continue this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/38#issuecomment-254975164,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkfbL6n2XU1CT3uAZ3ssSZW-NTbfIzks5q1rBcgaJpZM4KMohG
> .

pedrobaeza commented:
OK, thanks. Let me know any doubt.

mmalorni commented:
@ThomasBinsfeld @zakiuu I have someone that can work on this PR and we could complete it next week. We need to have this for a client. Do you want us to take over or do you want our help in another way. Will it be ready shortly if we just help? 
Thanks

zakmak88 committed:
[ADD] product_variant_configurator

This module has been ported from the odoomrp repo.
the module name in the odoo mrp repo was product_variants_no_automatic_creation
the following improvement where done while porting this module:

* Improve the readme file with a technical description on how the module should
  be used with other modules.
* Add attribute col=4 to group parent for a better desplay of product category
  for view.
* Remove unnecessary @api.multi before compute or onchange methods.
* Move product.configurator.attribute model to the new
  product_configurator_attribute.py file.
* Add unit test for product.configurator.attribute model that includes test for
  _compute_possible_value_ids and _compute_price_extra methods.
* Rename XML files in views and security folder as per OCA guidelines.
* Improve test coverage for product_configurator.
* Add unit test for purchase.order and purchase.order.line models.
* Separate the purchase.order and purchase.order.line models code into two files.
* Rename XML files as per OCA guidelines.
* product_variant_configurator: set product_tmpl_id in
  onchange_product_id_product_configurator_old_api and
  onchange_product_id_product_configurator methods for product.configurator model.
* Add unit test for onchange_product_tmpl_id and onchange_product_attribute_ids.

zakiuu commented:
@mmalorni Any help is very welcomed.
@pedrobaeza getting back to work on the PR today

pedrobaeza commented:
I have extracted correctly the commit history and merge your work in https://github.com/tecnativa/product-variant/tree/8.0-product_variant_configurator. @mmalorni is going to work in this also this week. Please coordinate both.

zakiuu commented:
@pedrobaeza  @mmalorni  I'll continue working on increasing the test coverage and find a better implementation for the product pricelist

pedrobaeza commented:
OK, cool, then just replace your branch with mine and you'll have all the work correctly credited (I have squashed your work in one commit for simplicity).

lmignon commented:
@pedrobaeza Thank you for your help! I try to clarify in my head what you expect with this PR. If I check the diff between this branch and yours, I can see that the module `purchase_product_variant` is not part of you branch. Am I right? Do you expect a replace of the content of the branch referenced in this pr with the one you provide and an other PR with `purchase_product_variant`?

lmignon commented:
@pedrobaeza reset done on your branch!

pedrobaeza commented:
Hi, @lmignon. Yeah, the idea is to finish this module in one PR (for not delaying more their inclusion), and then make another PR on top of this one with the purchase_product_configurator module. This way, we respect the implicit rule of 1 module = 1 PR, and it's also easier to follow commit history, as joining both is going to be a very long list (although I squashed a bit the history).

What do you think?

lmignon commented:
@pedrobaeza As you can see I've reset this branch on yours. In the same time I prepare an other one with the work done on `purchase_product_configurator` based on this one and keeping the commit history from odoomrp

pedrobaeza commented:
Cool! Let me know if you need help with anything, although I know that you're a master git-fu :wink:

lmignon commented:
@pedrobaeza split and history recovery done for `purchase_product_variants` in #43

zakiuu commented:
Yeah any help is welcome

On Thu, Oct 20, 2016, 17:52 Martin Malorni notifications@github.com wrote:

> @ThomasBinsfeld https://github.com/ThomasBinsfeld @zakiuu
> https://github.com/zakiuu I have someone that can work on this PR and
> we could complete it next week. We need to have this for a client. Do you
> want us to take over or do you want our help in another way. Will it be
> ready shortly if we just help?
> Thanks
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/38#issuecomment-255163406,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkfarRk4RQ36sGGIB5anHDEDBQOrIXks5q15xQgaJpZM4KMohG
> .

zakmak88 committed:
[FIX] product_variant configurator: Fix _price_rule_get_multi

Fix _price_rule_get_multi method inspired from v10 implementation
so the migration will be easier to v9 and v10

zakmak88 committed:
[FIX] product_variant configurator: Unit test for  _price_rule_get_multi

zakmak88 committed:
[FIX] product_variant_configurator: ad sequence to pricelist test

zakiuu commented:
@pedrobaeza I've fixed the issue in the pricelist I think the branch can be merged now

zakmak88 committed:
[FIX]product_variant_configurator: Remove duplicated xml files

ghost commented:
@zakiuu hey there - do you still need help regarding the test coverage for this PR (or is it actually ready)? If so, I may have some time to work on these, just tell me where I should focus.

zakiuu commented:
@sc-microcom it will be really good if you add some the test coverage is at 86%

mmalorni commented:
On the variant tab view "product.template.product.form" of the product form, there is a message that says that modifying the list of variants attributes will delete variants and loose data. Maybe this text should be reviewed since variants are not created automatically in this module.
"adding or deleting attributes
will delete and recreate existing variants and lead
to the loss of their possible customizations."
Replace by:
"Variants will not be created automatically. When created, they will be limited to the attributes and values entered here"

mmalorni commented:
The form view "product.product.form" doesn't work. When you want to create a variant, it displays the old Attribute values drop down and then the grid doesn't work. It is not used since when you only put a value in the grid for one of the attributes it give a warning that the product already exists.

mmalorni commented:
Looks like the rename of the views and model files in commit 58dd25 has not removed the views and models they were replacing. We now have 10 views instead of 5. Some are identical duplicates but some have differences. Which one is the right one?

zakiuu commented:
I did fix that

On Wed, Oct 26, 2016, 19:19 Martin Malorni notifications@github.com wrote:

> Looks like the rename of the views and model files in commit 58dd25 has
> not removed the views and models they were replacing. We now have 10 views
> instead of 5. Some are identical duplicates but some have differences.
> Which one is the right one?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/38#issuecomment-256433978,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkfWgkWtDHCshphQgbGICEmHKoKAA8ks5q35m8gaJpZM4KMohG
> .

mmalorni commented:
@zakiuu Maybe the manipulation that @lmignon or @pedrobaeza did to get back the commit history and to split in two the PR to keep the purchase part separate caused those changes to be lost but in this branch the files are back. I see that in the original branch (8.0-add_product_variant_configurator_original) I assume, the files are not there as you say but this is not the branch that is in the PR.

zakiuu commented:
I'll push it later on

On Wed, Oct 26, 2016, 19:31 Zakaria Makrelouf z.makrelouf@gmail.com wrote:

> I did fix that
> 
> On Wed, Oct 26, 2016, 19:19 Martin Malorni notifications@github.com
> wrote:
> 
> Looks like the rename of the views and model files in commit 58dd25 has
> not removed the views and models they were replacing. We now have 10 views
> instead of 5. Some are identical duplicates but some have differences.
> Which one is the right one?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/38#issuecomment-256433978,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkfWgkWtDHCshphQgbGICEmHKoKAA8ks5q35m8gaJpZM4KMohG
> .

mmalorni commented:
@zakiuu Hi, do you need help to fix the files issue? Do you know what could have caused the problem? Is it one of your commit that is missing? 
We will go ahead and add tests and do a PR on your branch if it is ok. You can then merge it to add the tests.

zakiuu commented:
I've just pushed the commit fix

On Thu, Oct 27, 2016, 16:42 Martin Malorni notifications@github.com wrote:

> @zakiuu https://github.com/zakiuu Hi, do you need help to fix the files
> issue? Do you know what could have caused the problem? Is it one of your
> commit that is missing?
> We will go ahead and add tests and do a PR on your branch if it is ok. You
> can then merge it to add the tests.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/OCA/product-variant/pull/38#issuecomment-256681308,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AANkfXo2C-AGFm48Wx2W41X9O7W5B9q8ks5q4MZJgaJpZM4KMohG
> .

pedrobaeza commented:
Is this working then?

zakiuu commented:
@pedrobaeza it is a yes for me

pedrobaeza commented:
@mmalorni, do you confirm also that your issues are solved?

mmalorni commented:
@pedrobaeza I will test today (5h later than you guys). We started working on the test coverage but we will be delayed. My dev called sick today.

Sent from my iPad
Martin Malorni
@:mmalorni@videotron.ca
@:Martin.malorni@hec.ca
H:(450) 962-9842
M:(514) 831-7037

> On Oct 28, 2016, at 5:52 AM, Pedro M. Baeza notifications@github.com wrote:
> 
> @mmalorni, do you confirm also that your issues are solved?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

mmalorni commented:
@pedrobaeza @zakiuu , the removal of the duplicate forms didn't resolve the issues. My comments stands. The drop down is still not removed and the grid doesn't work.

mmalorni commented:
@pedrobaeza Do we really need to get the coverage to 96% ? This seems high for code coverage considering that some parts cannot be tested. Is there à max % where the scripts stops asking for more coverage?
We added some tests on some files and got the % to 82% but it will be hard to get this file to 96%.

pedrobaeza commented:
Well, the more we have, the better, but if you find hard times doing that, the important thing is to get this working and merge it, and continue improving on later versions if possible.

mmalorni commented:
@pedrobaeza @zakiuu @lmignon We have done a first PR on the acsone branch with the additional tests. Can someone merge those tests to see what the result is in the OCA PR? We will work on fixing the code tomorrow to make the module functional.

sbidoul commented:
@mmalorni thanks for the effort with the tests. I'm personally quite happy with the test coverage. Of course we come from almost 100% in an almost empty repo, so being at 80% is good enough for me.

Note we are currently putting all our efforts on the 9.0 branch in #44. In particular we have significantly reworked then onchanges (in particular in https://github.com/OCA/product-variant/pull/44/commits/8594867917526bc5bd7798f945ddc37f31435416) so the attribute list remains consistent whatever the order in which the product template and/or product are changed.

So if you absolutely need to work on 8.0, I can happily merge your PR in here. Otherwise if you can work on 9.0 it's best to do your PR on top of #44 and when everything is clean there (we're almost done), backport the changes to 8.0, it will be less efforts.

mmalorni commented:
@Sbidoul We actually do not need the v8, we just didn¹t know that the work
on the v9 was progressing independently and faster than on the v8. We
actually need the v9 version which we were planning on tackling after. We
will help you finish PR #44 and see if we can backport in v8.

mmalorni commented:
We found that the module product_variant_csv_import in v8 is preventing the module from working. The issue is not apparent yet in v9 because the import module has not been converted yet.

sbidoul commented:
Closing. If there is interest for a 8.0 version, better to backport from 9.0.
